File: app/src/main/java/me/zhanghai/android/files/provider/document/DocumentFileSystemProvider.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.document;

import android.net.Uri;
import android.os.ParcelFileDescriptor;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.URI;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java8.nio.channels.FileChannel;
import java8.nio.channels.SeekableByteChannel;
import java8.nio.file.AccessDeniedException;
import java8.nio.file.AccessMode;
import java8.nio.file.CopyOption;
import java8.nio.file.DirectoryStream;
import java8.nio.file.FileAlreadyExistsException;
import java8.nio.file.FileStore;
import java8.nio.file.FileSystem;
import java8.nio.file.FileSystemAlreadyExistsException;
import java8.nio.file.FileSystemNotFoundException;
import java8.nio.file.LinkOption;
import java8.nio.file.OpenOption;
import java8.nio.file.Path;
import java8.nio.file.ProviderMismatchException;
import java8.nio.file.StandardOpenOption;
import java8.nio.file.attribute.BasicFileAttributes;
import java8.nio.file.attribute.FileAttribute;
import java8.nio.file.attribute.FileAttributeView;
import java8.nio.file.spi.FileSystemProvider;
import java9.util.Objects;
import java9.util.function.Consumer;
import me.zhanghai.android.files.file.MimeTypes;
import me.zhanghai.android.files.provider.common.AccessModes;
import me.zhanghai.android.files.provider.common.ByteString;
import me.zhanghai.android.files.provider.common.ByteStringPath;
import me.zhanghai.android.files.provider.common.ByteStringUriUtils;
import me.zhanghai.android.files.provider.common.CopyOptions;
import me.zhanghai.android.files.provider.common.MoreFileChannels;
import me.zhanghai.android.files.provider.common.OpenOptions;
import me.zhanghai.android.files.provider.common.PathListDirectoryStream;
import me.zhanghai.android.files.provider.common.PathObservable;
import me.zhanghai.android.files.provider.common.PathObservableProvider;
import me.zhanghai.android.files.provider.common.Searchable;
import me.zhanghai.android.files.provider.common.WalkFileTreeSearchable;
import me.zhanghai.android.files.provider.content.resolver.Resolver;
import me.zhanghai.android.files.provider.content.resolver.ResolverException;
import me.zhanghai.android.files.provider.document.resolver.DocumentResolver;

public class DocumentFileSystemProvider extends FileSystemProvider
        implements PathObservableProvider, Searchable {

    static final String SCHEME = "document";

    private static final ByteString HIDDEN_FILE_NAME_PREFIX = ByteString.fromString(".");

    private static DocumentFileSystemProvider sInstance;
    private static final Object sInstanceLock = new Object();

    @NonNull
    private final Map<Uri, DocumentFileSystem> mFileSystems = new HashMap<>();

    private final Object mLock = new Object();

    private DocumentFileSystemProvider() {}

    public static void install() {
        synchronized (sInstanceLock) {
            if (sInstance != null) {
                throw new IllegalStateException();
            }
            sInstance = new DocumentFileSystemProvider();
            FileSystemProvider.installProvider(sInstance);
        }
    }

    public static boolean isDocumentPath(@NonNull Path path) {
        Objects.requireNonNull(path);
        return path instanceof DocumentPath;
    }

    @NonNull
    public static Uri getDocumentUri(@NonNull Path path) throws IOException {
        DocumentPath documentPath = requireDocumentPath(path);
        try {
            return DocumentResolver.getDocumentUri(documentPath);
        } catch (ResolverException e) {
            throw e.toFileSystemException(path.toString());
        }
    }

    @NonNull
    public static Uri getTreeUri(@NonNull Path path) {
        DocumentPath documentPath = requireDocumentPath(path);
        return documentPath.getTreeUri();
    }

    @NonNull
    public static Path getRootPathForTreeUri(@NonNull Uri treeUri) {
        Objects.requireNonNull(treeUri);
        return getOrNewFileSystem(treeUri).getRootDirectory();
    }

    @NonNull
    static DocumentFileSystem getOrNewFileSystem(@NonNull Uri treeUri) {
        synchronized (sInstance.mLock) {
            DocumentFileSystem fileSystem = sInstance.mFileSystems.get(treeUri);
            if (fileSystem != null) {
                return fileSystem;
            }
            return sInstance.newFileSystemLocked(treeUri);
        }
    }

    void removeFileSystem(@NonNull DocumentFileSystem fileSystem) {
        synchronized (sInstance.mLock) {
            Uri treeUri = fileSystem.getTreeUri();
            sInstance.mFileSystems.remove(treeUri);
        }
    }

    @NonNull
    @Override
    public String getScheme() {
        return SCHEME;
    }

    @NonNull
    @Override
    public FileSystem newFileSystem(@NonNull URI uri, @NonNull Map<String, ?> env) {
        Objects.requireNonNull(uri);
        requireSameScheme(uri);
        Objects.requireNonNull(env);
        Uri treeUri = getTreeUriFromUri(uri);
        synchronized (mLock) {
            DocumentFileSystem fileSystem = mFileSystems.get(treeUri);
            if (fileSystem != null) {
                throw new FileSystemAlreadyExistsException(treeUri.toString());
            }
            return newFileSystemLocked(treeUri);
        }
    }

    @NonNull
    private DocumentFileSystem newFileSystemLocked(@NonNull Uri treeUri) {
        DocumentFileSystem fileSystem = new DocumentFileSystem(this, treeUri);
        mFileSystems.put(treeUri, fileSystem);
        return fileSystem;
    }

    @NonNull
    @Override
    public FileSystem getFileSystem(@NonNull URI uri) {
        Objects.requireNonNull(uri);
        requireSameScheme(uri);
        Uri treeUri = getTreeUriFromUri(uri);
        synchronized (mLock) {
            DocumentFileSystem fileSystem = mFileSystems.get(treeUri);
            if (fileSystem == null) {
                throw new FileSystemNotFoundException(treeUri.toString());
            }
            return fileSystem;
        }
    }

    @NonNull
    private static Uri getTreeUriFromUri(@NonNull URI uri) {
        ByteString schemeSpecificPart = ByteStringUriUtils.getDecodedSchemeSpecificPart(uri);
        if (schemeSpecificPart == null) {
            throw new IllegalArgumentException("URI must have a scheme specific part");
        }
        return Uri.parse(schemeSpecificPart.toString());
    }

    @NonNull
    @Override
    public Path getPath(@NonNull URI uri) {
        Objects.requireNonNull(uri);
        requireSameScheme(uri);
        Uri treeUri = getTreeUriFromUri(uri);
        ByteString fragment = ByteStringUriUtils.getDecodedFragment(uri);
        if (fragment == null) {
            throw new IllegalArgumentException("URI must have a fragment");
        }
        return getOrNewFileSystem(treeUri).getPath(fragment);
    }

    private static void requireSameScheme(@NonNull URI uri) {
        if (!Objects.equals(uri.getScheme(), SCHEME)) {
            throw new IllegalArgumentException("URI scheme must be \"" + SCHEME + "\"");
        }
    }

    @NonNull
    @Override
    public InputStream newInputStream(@NonNull Path file, @NonNull OpenOption... options)
            throws IOException {
        DocumentPath documentFile = requireDocumentPath(file);
        Objects.requireNonNull(options);
        Set<OpenOption> optionsSet = new HashSet<>(Arrays.asList(options));
        boolean hasCreate = optionsSet.remove(StandardOpenOption.CREATE);
        boolean hasCreateNew = optionsSet.remove(StandardOpenOption.CREATE_NEW);
        OpenOptions openOptions = OpenOptions.fromSet(optionsSet);
        if (openOptions.hasWrite()) {
            throw new UnsupportedOperationException(StandardOpenOption.WRITE.toString());
        }
        if (openOptions.hasAppend()) {
            throw new UnsupportedOperationException(StandardOpenOption.APPEND.toString());
        }
        String mode = DocumentOpenOptions.toMode(openOptions);
        if (hasCreate || hasCreateNew) {
            boolean exists = DocumentResolver.exists(documentFile);
            if (hasCreateNew && exists) {
                throw new FileAlreadyExistsException(file.toString());
            }
            if (!exists) {
                Uri uri;
                try {
                    // TODO: Allow passing in a mime type?
                    uri = DocumentResolver.create(documentFile, MimeTypes.GENERIC_MIME_TYPE);
                } catch (ResolverException e) {
                    throw e.toFileSystemException(file.toString());
                }
                try {
                    return Resolver.openInputStream(uri, mode);
                } catch (ResolverException e) {
                    throw e.toFileSystemException(uri.toString());
                }
            }
        }
        try {
            return DocumentResolver.openInputStream(documentFile, mode);
        } catch (ResolverException e) {
            throw e.toFileSystemException(file.toString());
        }
    }

    @NonNull
    @Override
    public OutputStream newOutputStream(@NonNull Path file, @NonNull OpenOption... options)
            throws IOException {
        DocumentPath documentFile = requireDocumentPath(file);
        Objects.requireNonNull(options);
        Set<OpenOption> optionsSet = new HashSet<>(Arrays.asList(options));
        if (optionsSet.isEmpty()) {
            optionsSet.add(StandardOpenOption.CREATE);
            optionsSet.add(StandardOpenOption.TRUNCATE_EXISTING);
        }
        optionsSet.add(StandardOpenOption.WRITE);
        boolean hasCreate = optionsSet.remove(StandardOpenOption.CREATE);
        boolean hasCreateNew = optionsSet.remove(StandardOpenOption.CREATE_NEW);
        OpenOptions openOptions = OpenOptions.fromSet(optionsSet);
        String mode = DocumentOpenOptions.toMode(openOptions);
        if (hasCreate || hasCreateNew) {
            boolean exists = DocumentResolver.exists(documentFile);
            if (hasCreateNew && exists) {
                throw new FileAlreadyExistsException(file.toString());
            }
            if (!exists) {
                Uri uri;
                try {
                    // TODO: Allow passing in a mime type?
                    uri = DocumentResolver.create(documentFile, MimeTypes.GENERIC_MIME_TYPE);
                } catch (ResolverException e) {
                    throw e.toFileSystemException(file.toString());
                }
                try {
                    return Resolver.openOutputStream(uri, mode);
                } catch (ResolverException e) {
                    throw e.toFileSystemException(uri.toString());
                }
            }
        }
        try {
            return DocumentResolver.openOutputStream(documentFile, mode);
        } catch (ResolverException e) {
            throw e.toFileSystemException(file.toString());
        }
    }

    @NonNull
    @Override
    public FileChannel newFileChannel(@NonNull Path file,
                                      @NonNull Set<? extends OpenOption> options,
                                      @NonNull FileAttribute<?>... attributes) throws IOException {
        DocumentPath documentFile = requireDocumentPath(file);
        Objects.requireNonNull(options);
        Objects.requireNonNull(attributes);
        boolean hasCreate = options.remove(StandardOpenOption.CREATE);
        boolean hasCreateNew = options.remove(StandardOpenOption.CREATE_NEW);
        OpenOptions openOptions = OpenOptions.fromSet(options);
        String mode = DocumentOpenOptions.toMode(openOptions);
        if (attributes.length > 0) {
            throw new UnsupportedOperationException(Arrays.toString(attributes));
        }
        ParcelFileDescriptor pfd = null;
        if (hasCreate || hasCreateNew) {
            boolean exists = DocumentResolver.exists(documentFile);
            if (hasCreateNew && exists) {
                throw new FileAlreadyExistsException(file.toString());
            }
            if (!exists) {
                Uri uri;
                try {
                    // TODO: Allow passing in a mime type?
                    uri = DocumentResolver.create(documentFile, MimeTypes.GENERIC_MIME_TYPE);
                } catch (ResolverException e) {
                    throw e.toFileSystemException(file.toString());
                }
                try {
                    pfd = Resolver.openParcelFileDescriptor(uri, mode);
                } catch (ResolverException e) {
                    throw e.toFileSystemException(uri.toString());
                }
            }
        }
        if (pfd == null) {
            try {
                pfd = DocumentResolver.openParcelFileDescriptor(documentFile, mode);
            } catch (ResolverException e) {
                throw e.toFileSystemException(file.toString());
            }
        }
        return MoreFileChannels.open(pfd, mode);
    }

    @NonNull
    @Override
    public SeekableByteChannel newByteChannel(@NonNull Path file,
                                              @NonNull Set<? extends OpenOption> options,
                                              @NonNull FileAttribute<?>... attributes)
            throws IOException {
        requireDocumentPath(file);
        Objects.requireNonNull(options);
        Objects.requireNonNull(attributes);
        return newFileChannel(file, options, attributes);
    }

    @NonNull
    @Override
    public DirectoryStream<Path> newDirectoryStream(
            @NonNull Path directory, @NonNull DirectoryStream.Filter<? super Path> filter)
            throws IOException {
        DocumentPath documentDirectory = requireDocumentPath(directory);
        Objects.requireNonNull(filter);
        List<Path> children;
        try {
            //noinspection unchecked
            children = (List<Path>) (List<?>) DocumentResolver.queryChildren(documentDirectory);
        } catch (ResolverException e) {
            throw e.toFileSystemException(directory.toString());
        }
        // TODO: Handle DocumentsContract.EXTRA_LOADING, EXTRA_INFO and EXTRA_ERROR.
        return new PathListDirectoryStream(children, filter);
    }

    @Override
    public void createDirectory(@NonNull Path directory, @NonNull FileAttribute<?>... attributes)
            throws IOException {
        DocumentPath documentDirectory = requireDocumentPath(directory);
        Objects.requireNonNull(attributes);
        if (attributes.length > 0) {
            throw new UnsupportedOperationException(Arrays.toString(attributes));
        }
        try {
            DocumentResolver.create(documentDirectory, MimeTypes.DIRECTORY_MIME_TYPE);
        } catch (ResolverException e) {
            throw e.toFileSystemException(directory.toString());
        }
    }

    @Override
    public void createSymbolicLink(@NonNull Path link, @NonNull Path target,
                                   @NonNull FileAttribute<?>... attributes) {
        requireDocumentPath(link);
        requireDocumentOrByteStringPath(target);
        Objects.requireNonNull(attributes);
        throw new UnsupportedOperationException();
    }

    @Override
    public void createLink(@NonNull Path link, @NonNull Path existing) {
        requireDocumentPath(link);
        requireDocumentPath(existing);
        throw new UnsupportedOperationException();
    }

    @Override
    public void delete(@NonNull Path path) throws IOException {
        DocumentPath documentPath = requireDocumentPath(path);
        try {
            DocumentResolver.remove(documentPath);
        } catch (ResolverException e) {
            throw e.toFileSystemException(path.toString());
        }
    }

    @NonNull
    @Override
    public Path readSymbolicLink(@NonNull Path link) {
        requireDocumentPath(link);
        throw new UnsupportedOperationException();
    }

    @Override
    public void copy(@NonNull Path source, @NonNull Path target, @NonNull CopyOption... options)
            throws IOException {
        DocumentPath documentSource = requireDocumentPath(source);
        DocumentPath documentTarget = requireDocumentPath(target);
        Objects.requireNonNull(options);
        CopyOptions copyOptions = CopyOptions.fromArray(options);
        DocumentCopyMove.copy(documentSource, documentTarget, copyOptions);
    }

    @Override
    public void move(@NonNull Path source, @NonNull Path target, @NonNull CopyOption... options)
            throws IOException {
        DocumentPath documentSource = requireDocumentPath(source);
        DocumentPath documentTarget = requireDocumentPath(target);
        Objects.requireNonNull(options);
        CopyOptions copyOptions = CopyOptions.fromArray(options);
        DocumentCopyMove.move(documentSource, documentTarget, copyOptions);
    }

    @Override
    public boolean isSameFile(@NonNull Path path, @NonNull Path path2) {
        requireDocumentPath(path);
        Objects.requireNonNull(path2);
        // TODO: DocumentsContract.findDocumentPath()?
        return Objects.equals(path, path2);
    }

    @Override
    public boolean isHidden(@NonNull Path path) {
        DocumentPath documentPath = requireDocumentPath(path);
        DocumentPath fileName = documentPath.getFileName();
        if (fileName == null) {
            return false;
        }
        ByteString fileNameBytes = fileName.toByteString();
        return fileNameBytes.startsWith(HIDDEN_FILE_NAME_PREFIX);
    }

    @NonNull
    @Override
    public FileStore getFileStore(@NonNull Path path) {
        requireDocumentPath(path);
        throw new UnsupportedOperationException();
    }

    @Override
    public void checkAccess(@NonNull Path path, @NonNull AccessMode... modes) throws IOException {
        DocumentPath documentPath = requireDocumentPath(path);
        Objects.requireNonNull(modes);
        AccessModes accessModes = AccessModes.fromArray(modes);
        if (accessModes.hasExecute()) {
            throw new AccessDeniedException(path.toString());
        }
        if (accessModes.hasWrite()) {
            try (OutputStream outputStream = DocumentResolver.openOutputStream(documentPath, "w")) {
                // Do nothing.
            } catch (ResolverException e) {
                throw e.toFileSystemException(path.toString());
            }
        }
        if (accessModes.hasRead()) {
            try (InputStream inputStream = DocumentResolver.openInputStream(documentPath, "r")) {
                // Do nothing.
            } catch (ResolverException e) {
                throw e.toFileSystemException(path.toString());
            }
        }
        if (!(accessModes.hasRead() || accessModes.hasWrite())) {
            try {
                DocumentResolver.checkExistence(documentPath);
            } catch (ResolverException e) {
                throw e.toFileSystemException(path.toString());
            }
        }
    }

    @Nullable
    @Override
    public <V extends FileAttributeView> V getFileAttributeView(@NonNull Path path,
                                                                @NonNull Class<V> type,
                                                                @NonNull LinkOption... options) {
        requireDocumentPath(path);
        Objects.requireNonNull(type);
        Objects.requireNonNull(options);
        if (!supportsFileAttributeView(type)) {
            return null;
        }
        //noinspection unchecked
        return (V) getFileAttributeView(path);
    }

    static boolean supportsFileAttributeView(@NonNull Class<? extends FileAttributeView> type) {
        return type.isAssignableFrom(DocumentFileAttributeView.class);
    }

    @NonNull
    @Override
    public <A extends BasicFileAttributes> A readAttributes(@NonNull Path path,
                                                            @NonNull Class<A> type,
                                                            @NonNull LinkOption... options)
            throws IOException {
        requireDocumentPath(path);
        Objects.requireNonNull(type);
        Objects.requireNonNull(options);
        if (!type.isAssignableFrom(DocumentFileAttributes.class)) {
            throw new UnsupportedOperationException(type.toString());
        }
        //noinspection unchecked
        return (A) getFileAttributeView(path).readAttributes();
    }

    private static DocumentFileAttributeView getFileAttributeView(@NonNull Path path) {
        DocumentPath documentPath = requireDocumentPath(path);
        return new DocumentFileAttributeView(documentPath);
    }

    @NonNull
    @Override
    public Map<String, Object> readAttributes(@NonNull Path path, @NonNull String attributes,
                                              @NonNull LinkOption... options) {
        requireDocumentPath(path);
        Objects.requireNonNull(attributes);
        Objects.requireNonNull(options);
        throw new UnsupportedOperationException();
    }

    @Override
    public void setAttribute(@NonNull Path path, @NonNull String attribute, @NonNull Object value,
                             @NonNull LinkOption... options) {
        requireDocumentPath(path);
        Objects.requireNonNull(attribute);
        Objects.requireNonNull(value);
        Objects.requireNonNull(options);
        throw new UnsupportedOperationException();
    }

    @NonNull
    @Override
    public PathObservable observePath(@NonNull Path path, long intervalMillis) throws IOException {
        DocumentPath documentDirectory = requireDocumentPath(path);
        return new DocumentPathObservable(documentDirectory, intervalMillis);
    }

    @Override
    public void search(@NonNull Path directory, @NonNull String query,
                       @NonNull Consumer<List<Path>> listener, long intervalMillis)
            throws IOException {
        requireDocumentPath(directory);
        Objects.requireNonNull(query);
        Objects.requireNonNull(listener);
        WalkFileTreeSearchable.search(directory, query, listener, intervalMillis);
    }

    private static DocumentPath requireDocumentPath(@NonNull Path path) {
        Objects.requireNonNull(path);
        if (!(path instanceof DocumentPath)) {
            throw new ProviderMismatchException(path.toString());
        }
        return (DocumentPath) path;
    }

    private static ByteString requireDocumentOrByteStringPath(@NonNull Path path) {
        Objects.requireNonNull(path);
        if (path instanceof DocumentPath) {
            return ((DocumentPath) path).toByteString();
        } else if (path instanceof ByteStringPath) {
            return ((ByteStringPath) path).toByteString();
        } else {
            throw new ProviderMismatchException(path.toString());
        }
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/document/DocumentFileSystemProvider.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.document

import android.net.Uri
import android.os.Build
import android.os.ParcelFileDescriptor
import androidx.core.net.toUri
import java8.nio.channels.FileChannel
import java8.nio.channels.SeekableByteChannel
import java8.nio.file.AccessDeniedException
import java8.nio.file.AccessMode
import java8.nio.file.CopyOption
import java8.nio.file.DirectoryStream
import java8.nio.file.FileAlreadyExistsException
import java8.nio.file.FileStore
import java8.nio.file.FileSystem
import java8.nio.file.FileSystemAlreadyExistsException
import java8.nio.file.FileSystemNotFoundException
import java8.nio.file.LinkOption
import java8.nio.file.OpenOption
import java8.nio.file.Path
import java8.nio.file.ProviderMismatchException
import java8.nio.file.StandardOpenOption
import java8.nio.file.attribute.BasicFileAttributes
import java8.nio.file.attribute.FileAttribute
import java8.nio.file.attribute.FileAttributeView
import java8.nio.file.spi.FileSystemProvider
import me.zhanghai.android.files.file.MimeType
import me.zhanghai.android.files.provider.common.ByteStringPath
import me.zhanghai.android.files.provider.common.PathListDirectoryStream
import me.zhanghai.android.files.provider.common.PathObservable
import me.zhanghai.android.files.provider.common.PathObservableProvider
import me.zhanghai.android.files.provider.common.Searchable
import me.zhanghai.android.files.provider.common.WalkFileTreeSearchable
import me.zhanghai.android.files.provider.common.decodedPathByteString
import me.zhanghai.android.files.provider.common.decodedQueryByteString
import me.zhanghai.android.files.provider.common.open
import me.zhanghai.android.files.provider.common.toAccessModes
import me.zhanghai.android.files.provider.common.toByteString
import me.zhanghai.android.files.provider.common.toCopyOptions
import me.zhanghai.android.files.provider.common.toOpenOptions
import me.zhanghai.android.files.provider.content.resolver.Resolver
import me.zhanghai.android.files.provider.content.resolver.ResolverException
import me.zhanghai.android.files.provider.document.resolver.DocumentResolver
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.URI

object DocumentFileSystemProvider : FileSystemProvider(), PathObservableProvider, Searchable {
    private const val SCHEME = "document"

    private val HIDDEN_FILE_NAME_PREFIX = ".".toByteString()

    private val fileSystems = mutableMapOf<Uri, DocumentFileSystem>()

    private val lock = Any()

    override fun getScheme(): String = SCHEME

    override fun newFileSystem(uri: URI, env: Map<String, *>): FileSystem {
        uri.requireSameScheme()
        val treeUri = uri.treeUri
        synchronized(lock) {
            if (fileSystems[treeUri] != null) {
                throw FileSystemAlreadyExistsException(treeUri.toString())
            }
            return newFileSystemLocked(treeUri)
        }
    }

    internal fun getOrNewFileSystem(treeUri: Uri): DocumentFileSystem =
        synchronized(lock) { fileSystems[treeUri] ?: newFileSystemLocked(treeUri) }

    private fun newFileSystemLocked(treeUri: Uri): DocumentFileSystem {
        val fileSystem = DocumentFileSystem(this, treeUri)
        fileSystems[treeUri] = fileSystem
        return fileSystem
    }

    override fun getFileSystem(uri: URI): FileSystem {
        uri.requireSameScheme()
        val treeUri = uri.treeUri
        return synchronized(lock) { fileSystems[treeUri] }
            ?: throw FileSystemNotFoundException(treeUri.toString())
    }

    internal fun removeFileSystem(fileSystem: DocumentFileSystem) {
        val treeUri = fileSystem.treeUri
        synchronized(lock) { fileSystems.remove(treeUri) }
    }

    override fun getPath(uri: URI): Path {
        uri.requireSameScheme()
        val treeUri = uri.treeUri
        val path = uri.decodedQueryByteString
            ?: throw IllegalArgumentException("URI must have a query")
        return getOrNewFileSystem(treeUri).getPath(path)
    }

    private fun URI.requireSameScheme() {
        val scheme = scheme
        require(scheme == SCHEME) { "URI scheme $scheme must be $SCHEME" }
    }

    private val URI.treeUri: Uri
        get() {
            val path = decodedPathByteString
                ?: throw IllegalArgumentException("URI must have a path")
            // Drop the first character which is always a slash.
            return path.toString().drop(1).toUri()
        }

    @Throws(IOException::class)
    override fun newInputStream(file: Path, vararg options: OpenOption): InputStream {
        file as? DocumentPath ?: throw ProviderMismatchException(file.toString())
        val optionsSet = mutableSetOf(*options)
        val create = optionsSet.remove(StandardOpenOption.CREATE)
        val createNew = optionsSet.remove(StandardOpenOption.CREATE_NEW)
        val openOptions = optionsSet.toOpenOptions()
        if (openOptions.write) {
            throw UnsupportedOperationException(StandardOpenOption.WRITE.toString())
        }
        if (openOptions.append) {
            throw UnsupportedOperationException(StandardOpenOption.APPEND.toString())
        }
        val mode = openOptions.toDocumentMode()
        if (create || createNew) {
            val exists = DocumentResolver.exists(file)
            if (createNew && exists) {
                throw FileAlreadyExistsException(file.toString())
            }
            if (!exists) {
                val uri = try {
                    // TODO: Allow passing in a mime type?
                    DocumentResolver.create(file, MimeType.GENERIC.value)
                } catch (e: ResolverException) {
                    throw e.toFileSystemException(file.toString())
                }
                return try {
                    Resolver.openInputStream(uri, mode)
                } catch (e: ResolverException) {
                    throw e.toFileSystemException(uri.toString())
                }
            }
        }
        return try {
            DocumentResolver.openInputStream(file, mode)
        } catch (e: ResolverException) {
            throw e.toFileSystemException(file.toString())
        }
    }

    @Throws(IOException::class)
    override fun newOutputStream(file: Path, vararg options: OpenOption): OutputStream {
        file as? DocumentPath ?: throw ProviderMismatchException(file.toString())
        val optionsSet = mutableSetOf(*options)
        if (optionsSet.isEmpty()) {
            optionsSet += StandardOpenOption.CREATE
            optionsSet += StandardOpenOption.TRUNCATE_EXISTING
        }
        optionsSet += StandardOpenOption.WRITE
        val create = optionsSet.remove(StandardOpenOption.CREATE)
        val createNew = optionsSet.remove(StandardOpenOption.CREATE_NEW)
        val openOptions = optionsSet.toOpenOptions()
        val mode = openOptions.toDocumentMode()
        if (create || createNew) {
            val exists = DocumentResolver.exists(file)
            if (createNew && exists) {
                throw FileAlreadyExistsException(file.toString())
            }
            if (!exists) {
                val uri = try {
                    // TODO: Allow passing in a mime type?
                    DocumentResolver.create(file, MimeType.GENERIC.value)
                } catch (e: ResolverException) {
                    throw e.toFileSystemException(file.toString())
                }
                return try {
                    Resolver.openOutputStream(uri, mode)
                } catch (e: ResolverException) {
                    throw e.toFileSystemException(uri.toString())
                }
            }
        }
        return try {
            DocumentResolver.openOutputStream(file, mode)
        } catch (e: ResolverException) {
            throw e.toFileSystemException(file.toString())
        }
    }

    @Throws(IOException::class)
    override fun newFileChannel(
        file: Path,
        options: Set<OpenOption>,
        vararg attributes: FileAttribute<*>
    ): FileChannel {
        file as? DocumentPath ?: throw ProviderMismatchException(file.toString())
        val options = options.toMutableSet()
        val hasCreate = options.remove(StandardOpenOption.CREATE)
        val hasCreateNew = options.remove(StandardOpenOption.CREATE_NEW)
        val openOptions = options.toOpenOptions()
        val mode = openOptions.toDocumentMode()
        if (attributes.isNotEmpty()) {
            throw UnsupportedOperationException(attributes.contentToString())
        }
        var pfd: ParcelFileDescriptor? = null
        if (hasCreate || hasCreateNew) {
            val exists = DocumentResolver.exists(file)
            if (hasCreateNew && exists) {
                throw FileAlreadyExistsException(file.toString())
            }
            if (!exists) {
                val uri = try {
                    // TODO: Allow passing in a mime type?
                    DocumentResolver.create(file, MimeType.GENERIC.value)
                } catch (e: ResolverException) {
                    throw e.toFileSystemException(file.toString())
                }
                pfd = try {
                    Resolver.openParcelFileDescriptor(uri, mode)
                } catch (e: ResolverException) {
                    throw e.toFileSystemException(uri.toString())
                }
            }
        }
        if (pfd == null) {
            pfd = try {
                DocumentResolver.openParcelFileDescriptor(file, mode)
            } catch (e: ResolverException) {
                throw e.toFileSystemException(file.toString())
            }
        }
        // TODO: kotlinc: Type mismatch: inferred type is ParcelFileDescriptor? but
        //  ParcelFileDescriptor was expected
        //return FileChannel::class.open(pfd, mode)
        return FileChannel::class.open(pfd!!, mode)
    }

    @Throws(IOException::class)
    override fun newByteChannel(
        file: Path,
        options: Set<OpenOption>,
        vararg attributes: FileAttribute<*>
    ): SeekableByteChannel {
        file as? DocumentPath ?: throw ProviderMismatchException(file.toString())
        return newFileChannel(file, options, *attributes)
    }

    @Throws(IOException::class)
    override fun newDirectoryStream(
        directory: Path,
        filter: DirectoryStream.Filter<in Path>
    ): DirectoryStream<Path> {
        directory as? DocumentPath ?: throw ProviderMismatchException(directory.toString())
        val children = try {
            @Suppress("UNCHECKED_CAST")
            DocumentResolver.queryChildren(directory) as List<Path>
        } catch (e: ResolverException) {
            throw e.toFileSystemException(directory.toString())
        }
        // TODO: Handle DocumentsContract.EXTRA_LOADING, EXTRA_INFO and EXTRA_ERROR.
        return PathListDirectoryStream(children, filter)
    }

    @Throws(IOException::class)
    override fun createDirectory(directory: Path, vararg attributes: FileAttribute<*>) {
        directory as? DocumentPath ?: throw ProviderMismatchException(directory.toString())
        if (attributes.isNotEmpty()) {
            throw UnsupportedOperationException(attributes.contentToString())
        }
        try {
            DocumentResolver.create(directory, MimeType.DIRECTORY.value)
        } catch (e: ResolverException) {
            throw e.toFileSystemException(directory.toString())
        }
    }

    override fun createSymbolicLink(link: Path, target: Path, vararg attributes: FileAttribute<*>) {
        link as? DocumentPath ?: throw ProviderMismatchException(link.toString())
        when (target) {
            is DocumentPath, is ByteStringPath -> {}
            else -> throw ProviderMismatchException(target.toString())
        }
        throw UnsupportedOperationException()
    }

    override fun createLink(link: Path, existing: Path) {
        link as? DocumentPath ?: throw ProviderMismatchException(link.toString())
        existing as? DocumentPath ?: throw ProviderMismatchException(existing.toString())
        throw UnsupportedOperationException()
    }

    @Throws(IOException::class)
    override fun delete(path: Path) {
        path as? DocumentPath ?: throw ProviderMismatchException(path.toString())
        try {
            DocumentResolver.remove(path)
        } catch (e: ResolverException) {
            throw e.toFileSystemException(path.toString())
        }
    }

    override fun readSymbolicLink(link: Path): Path {
        link as? DocumentPath ?: throw ProviderMismatchException(link.toString())
        throw UnsupportedOperationException()
    }

    @Throws(IOException::class)
    override fun copy(source: Path, target: Path, vararg options: CopyOption) {
        source as? DocumentPath ?: throw ProviderMismatchException(source.toString())
        target as? DocumentPath ?: throw ProviderMismatchException(target.toString())
        val copyOptions = options.toCopyOptions()
        DocumentCopyMove.copy(source, target, copyOptions)
    }

    @Throws(IOException::class)
    override fun move(source: Path, target: Path, vararg options: CopyOption) {
        source as? DocumentPath ?: throw ProviderMismatchException(source.toString())
        target as? DocumentPath ?: throw ProviderMismatchException(target.toString())
        val copyOptions = options.toCopyOptions()
        DocumentCopyMove.move(source, target, copyOptions)
    }

    override fun isSameFile(path: Path, path2: Path): Boolean {
        path as? DocumentPath ?: throw ProviderMismatchException(path.toString())
        // TODO: DocumentsContract.findDocumentPath()?
        return path == path2
    }

    override fun isHidden(path: Path): Boolean {
        path as? DocumentPath ?: throw ProviderMismatchException(path.toString())
        val fileName = path.fileNameByteString ?: return false
        return fileName.startsWith(HIDDEN_FILE_NAME_PREFIX)
    }

    override fun getFileStore(path: Path): FileStore {
        path as? DocumentPath ?: throw ProviderMismatchException(path.toString())
        throw UnsupportedOperationException()
    }

    @Throws(IOException::class)
    override fun checkAccess(path: Path, vararg modes: AccessMode) {
        path as? DocumentPath ?: throw ProviderMismatchException(path.toString())
        // This checks existence as well.
        val mimeType = try {
            DocumentResolver.getMimeType(path)
        } catch (e: ResolverException) {
            throw e.toFileSystemException(path.toString())
        }
        val isDirectory = mimeType == MimeType.DIRECTORY.value
        if (isDirectory) {
            // There's no elegant way to check access to a directory beyond its existence.
            return
        }
        val accessModes = modes.toAccessModes()
        if (accessModes.execute) {
            throw AccessDeniedException(path.toString())
        }
        if (accessModes.write) {
            // Before Android 10, ParcelFileDescriptor.parseMode() parses "w" as "wt", and we would
            // truncate the file to empty. So work around that with "wa" on older platforms.
            val mode = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) "w" else "wa"
            try {
                DocumentResolver.openOutputStream(path, mode).use {}
            } catch (e: ResolverException) {
                throw e.toFileSystemException(path.toString())
            }
        }
        if (accessModes.read) {
            try {
                DocumentResolver.openInputStream(path, "r").use {}
            } catch (e: ResolverException) {
                throw e.toFileSystemException(path.toString())
            }
        }
    }

    override fun <V : FileAttributeView> getFileAttributeView(
        path: Path,
        type: Class<V>,
        vararg options: LinkOption
    ): V? {
        if (!supportsFileAttributeView(type)) {
            return null
        }
        @Suppress("UNCHECKED_CAST")
        return getFileAttributeView(path) as V
    }

    internal fun supportsFileAttributeView(type: Class<out FileAttributeView>): Boolean =
        type.isAssignableFrom(DocumentFileAttributeView::class.java)

    @Throws(IOException::class)
    override fun <A : BasicFileAttributes> readAttributes(
        path: Path,
        type: Class<A>,
        vararg options: LinkOption
    ): A {
        if (!type.isAssignableFrom(DocumentFileAttributes::class.java)) {
            throw UnsupportedOperationException(type.toString())
        }
        @Suppress("UNCHECKED_CAST")
        return getFileAttributeView(path).readAttributes() as A
    }

    private fun getFileAttributeView(path: Path): DocumentFileAttributeView {
        path as? DocumentPath ?: throw ProviderMismatchException(path.toString())
        return DocumentFileAttributeView(path)
    }

    override fun readAttributes(
        path: Path,
        attributes: String,
        vararg options: LinkOption
    ): Map<String, Any> {
        path as? DocumentPath ?: throw ProviderMismatchException(path.toString())
        throw UnsupportedOperationException()
    }

    override fun setAttribute(
        path: Path,
        attribute: String,
        value: Any,
        vararg options: LinkOption
    ) {
        path as? DocumentPath ?: throw ProviderMismatchException(path.toString())
        throw UnsupportedOperationException()
    }

    @Throws(IOException::class)
    override fun observe(path: Path, intervalMillis: Long): PathObservable {
        path as? DocumentPath ?: throw ProviderMismatchException(path.toString())
        return DocumentPathObservable(path, intervalMillis)
    }

    @Throws(IOException::class)
    override fun search(
        directory: Path,
        query: String,
        intervalMillis: Long,
        listener: (List<Path>) -> Unit
    ) {
        directory as? DocumentPath ?: throw ProviderMismatchException(directory.toString())
        WalkFileTreeSearchable.search(directory, query, intervalMillis, listener)
    }
}
